### PR TITLE
Add empty bodies for added CPU API funtions, modify MPI initializaiton

### PR DESCRIPTION
**Pull request by huwzpf**
Depends on #15 #16

### CODE REVIEW

**Review by MAJ0RRR — APPROVED** [review]

File: cudampilib/include/cudampicommon.h [hunk 1]
```diff
@@ -11,7 +11,10 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OU
 */
 #include <stdio.h>
 #include <stdlib.h>
+#include <cuda_runtime.h>
 
 float computeDevPerformance(struct timeval period);
 
 float getGPUpower(int gpuid);
+
+cudaError_t __cudampi__getCpuFreeThreads(int* count)
```
Comment (Kepins) [inline cudampilib/include/cudampicommon.h:20]: missing semicolon

Comment (Kepins) [inline cudampilib/include/cudampicommon.h:20]: missing semicolon

File: cudampilib/include/cudampicommon.h [hunk 2]
```diff
@@ -11,7 +11,10 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OU
 */
 #include <stdio.h>
 #include <stdlib.h>
+#include <cuda_runtime.h>
```
Comment (Kepins) [inline cudampilib/include/cudampicommon.h:14]: I believe if we want to include `<cuda_runtime.h>` in this file we need to modify our `compile` script to include this on line 13:
`gcc -c cudampicommon.c -Iinclude -o $BUILD_DIR/cudampicommon.o -I/usr/local/cuda/include`

File: cudampilib/cudampilib.c [hunk 3]
```diff
@@ -758,6 +845,17 @@ void __cudampi__kernel(void *devPtr) {
   }
 }
 
+void cpukernel(void *devPtr, int num_threads);
```
Comment (Kepins) [inline cudampilib/cudampilib.c:848]: This function declaration should come from `apps/APPLICATION_NAME/appcpukernel` or similar. Maybe we should include empty definition for each cpu kernel (for each app) so there would be no errors:

```
/usr/bin/ld: build/cudampilib.o: in function `__cudampi__cpuKernel':
cudampilib.c:(.text+0x3337): undefined reference to `cpukernel'
```
during compilation of any app.

File: cudampilib/cudampilib.c [hunk 4]
```diff
@@ -710,6 +793,10 @@ cudaError_t __cudampi__cudaMemcpyAsync(void *dst, const void *src, size_t count,
   }
 }
 
+cudaError_t __cudampi__cpuMemcpyAsync(void *dst, const void *src, size_t count, enum cudaMemcpyKind kind) {
+  return cudaSuccess;
+}
```
Comment (Kepins) [inline cudampilib/cudampilib.c:798]: Include TODO with appropriate comment so we don't miss it later on or remove this function all together.

File: cudampilib/cudampilib.c [hunk 5]
```diff
@@ -530,7 +590,10 @@ cudaError_t __cudampi__deviceSynchronize(void) {
     }
   }
 
-  if (__cudampi_isLocalGpu) { // run locally
+  if (__cudampi__isCpu()) { // run CPU synchronization remotely
+    // wyjebać local cpu
```
Comment (Kepins) [inline cudampilib/cudampilib.c:594]: There is a possibility we will be showing this code to Czarnul. xD

File: cudampilib/cudampilib.c [hunk 6]
```diff
@@ -21,16 +21,18 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OU
 
 // int __cudampi__GPUcountpernode=1;
 
-
 #define __cudampi__currentDevice  __cudampi__currentdevice[omp_get_thread_num()]
 #define __cudampi__currentCommunicator  __cudampi__communicators[__cudampi__currentDevice]
-
 #define  __cudampi_isLocalGpu __cudampi__currentDevice < __cudampi__GPUcountspernode[0]
```
Comment (Kepins) [inline cudampilib/cudampilib.c:26]: nit. Two spaces after `#define`

File: cudampilib/cudampilib.c [hunk 7]
```diff
@@ -616,6 +679,21 @@ cudaError_t __cudampi__cudaSetDevice(int device) {
   }
 }
 
+int __cudampi__isCpu()
+{
+  // TODO
+  return __cudampi__currentDevice  >= __cudampi_totalgpudevicecount;
+}
```
Comment (Kepins) [inline cudampilib/cudampilib.c:686]: We have such define for checking if isLocalGpu.
`#define  __cudampi_isLocalGpu __cudampi__currentDevice < __cudampi__GPUcountspernode[0]`
Couldn't this functionality also be a define?

File: cudampilib/cudampilib.c [hunk 8]
```diff
@@ -758,6 +845,17 @@ void __cudampi__kernel(void *devPtr) {
   }
 }
 
+void cpukernel(void *devPtr, int num_threads);
+
+void __cudampi__cpuKernel(void *devPtr){
+
+  if (__cudampi__isLocalCpu) { // run locally
```
Comment (Kepins) [inline cudampilib/cudampilib.c:852]: `__cudampi__isLocalCpu` is not defined maybe at least add it TODO?

**Review by Kepins — CHANGES_REQUESTED** [review]

File: cudampilib/include/cudampicommon.h (hunk 2 above)
Comment (huwzpf) [reply cudampilib/include/cudampicommon.h:14]: Compilation will be fixed in #13.

File: cudampilib/cudampilib.c (hunk 3 above)
Comment (huwzpf) [reply cudampilib/cudampilib.c:848]: The function will be defined in collatz app (and compilation of other apps disabled) in #13

File: cudampilib/cudampilib.c (hunk 8 above)
Comment (huwzpf) [reply cudampilib/cudampilib.c:853]: As discussed, we won't use local CPUs. It should be removed.

**Pull request merged**